test/cpp/util/spiffe_test_credentials.cc

```diff
@@ -0,0 +1,138 @@
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    arg->set_success(1);
+    arg->set_status(GRPC_STATUS_OK);
+    return 0;
+  }
+};
+
+static void TestAsyncTlsServerAuthorizationCheckCallback(
+    ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) {
+  GPR_ASSERT(arg != nullptr);
+  arg->set_success(1);
+  arg->set_status(GRPC_STATUS_OK);
+  arg->OnServerAuthorizationCheckDoneCallback();
+}
+
+class TestAsyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+ public:
+  TestAsyncTlsServerAuthorizationCheck(SpiffeThreadList* thread_list)
+      : thread_list_(thread_list) {}
+
+  // Async implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    server_authz_check_thread_ =
+        std::thread(TestAsyncTlsServerAuthorizationCheckCallback, arg);
+    thread_list_->add_thread(std::move(server_authz_check_thread_));
+    return 1;
+  }
+
+ private:
+  std::thread server_authz_check_thread_;
+  SpiffeThreadList* thread_list_;
+};
+
+/** This method creates a TlsCredentialsOptions instance with no key materials,
+ *  whose credential reload config is configured using the
+ *  TestSyncTlsCredentialReload class, and whose server authorization check
+ *  config is determined as follows:
+ *  - if |is_client| is true,
+ *      -if |is_async|, configured by TestAsyncTlsServerAuthorizationCheck,
+ *      -otherwise, configured by TestSyncTlsServerAuthorizationCheck.
+ *  - otherwise, the server authorization check config is not populated.
+ *
+ *  Further, the cert request type of the options instance is always set to
+ *  GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY for both the
+ *  client and the server. **/
+std::shared_ptr<::grpc_impl::experimental::TlsCredentialsOptions>
+CreateTestTlsCredentialsOptions(bool is_client, bool is_async,
+                                SpiffeThreadList* thread_list) {
+  /** Create a credential reload config that is configured using the
+   *  TestSyncTlsCredentialReload class. **/
+  std::shared_ptr<TestSyncTlsCredentialReload> credential_reload_interface(
+      new TestSyncTlsCredentialReload());
+  std::shared_ptr<::grpc_impl::experimental::TlsCredentialReloadConfig>
+      test_credential_reload_config(
+          new ::grpc_impl::experimental::TlsCredentialReloadConfig(
+              credential_reload_interface));
+
+  /** Create a server authorization check config that is configured depending on
+   *  the parameters |is_client| and |is_async|. **/
+  std::shared_ptr<::grpc_impl::experimental::TlsServerAuthorizationCheckConfig>
+      test_server_authorization_check_config = nullptr;
+  if (is_client) {
```

Comment:
Can we return early if `is_client` is `false` to reduce the nesting?

Reply:
Done.